### PR TITLE
[improve][txn] Support ack message list for transaction

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation
Now, there is
```java
doAcknowledgeWithTxn(List<MessageId> messageIdList, AckType ackType,
                                                           Map<String, Long> properties, TransactionImpl txn)
```
But not interface
```java
acknowledgeAsync(List<MessageId> messageIdList, Transaction txn)
```
```java
CompletableFuture<Void> acknowledgeAsync(Messages<?> messages, Transaction txn);
```
### Modification
Add interface
```java
acknowledgeAsync(List<MessageId> messageIdList, Transaction txn)
```
```java
CompletableFuture<Void> acknowledgeAsync(Messages<?> messages, Transaction txn);
```

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)
- [x] `doc-not-needed`

- [x] `doc-label-missing`

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java [hunk 1]
```diff
@@ -997,4 +998,47 @@ public void testConsistencyOfTransactionStatsAtEndTxn() throws Exception {
 
         transaction.commit().get();
     }
+
+    @Test
+    public void testTransactionAckMessageList() throws Exception {
+        String topic = "persistent://" + NAMESPACE1 +"/test";
+        String subName = "testSub";
+
+        Producer<byte[]> producer = pulsarClient.newProducer()
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1007]: Please add the annotation `@Cleanup` for the producer.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java [hunk 2]
```diff
@@ -997,4 +998,47 @@ public void testConsistencyOfTransactionStatsAtEndTxn() throws Exception {
...
+                .newTransaction()
+                .withTransactionTimeout(5, TimeUnit.MINUTES)
+                .build()
+                .get();
+
+        consumer.acknowledgeAsync(messages, transaction);
+        transaction.commit().get();
+
+        consumer.close();
+        consumer = pulsarClient.newConsumer()
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1034]: Please close the consumer finally.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java [hunk 3]
```diff
@@ -997,4 +998,47 @@ public void testConsistencyOfTransactionStatsAtEndTxn() throws Exception {
 
         transaction.commit().get();
     }
+
+    @Test
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1002]: Please add a timeout param for the test.

**Review by codelipenghui — COMMENTED** [review]
Please add support for `CompletableFuture<Void> acknowledgeAsync(Messages<?> messages,  Transaction txn);`

**Review by codelipenghui — COMMENTED** [review]
Please add support for `CompletableFuture<Void> acknowledgeAsync(Messages<?> messages,  Transaction txn);`

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java [hunk 4]
```diff
@@ -491,13 +491,51 @@ CompletableFuture<Void> acknowledgeCumulativeAsync(MessageId messageId,
...
+     *             if the consumer was already closed
+     * @throws org.apache.pulsar.client.api.PulsarClientException.TransactionConflictException
+     *             if the ack with messageId is less than the messageId in pending ack state or ack with transaction is
+     *             different from the transaction in pending ack.
+     * @throws org.apache.pulsar.client.api.PulsarClientException.NotAllowedException
+     *             broker don't support transaction
+     * @return {@link CompletableFuture} the future of the ack result
+     * */
+    CompletableFuture<Void> acknowledgeAsync(Messages<?> messages, Transaction txn) throws PulsarClientException;
+
```
Comment (Technoboy-) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java:512]: It's better not throw exception when using CompletableFuture as return type.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java [hunk 5]
```diff
@@ -491,13 +491,51 @@ CompletableFuture<Void> acknowledgeCumulativeAsync(MessageId messageId,
...
+     * @throws PulsarClientException.AlreadyClosedException
+     *             if the consumer was already closed
+     * @throws org.apache.pulsar.client.api.PulsarClientException.TransactionConflictException
+     *             if the ack with messageId is less than the messageId in pending ack state or ack with transaction is
+     *             different from the transaction in pending ack.
+     * @throws org.apache.pulsar.client.api.PulsarClientException.NotAllowedException
+     *             broker don't support transaction
+     * @return {@link CompletableFuture} the future of the ack result     */
+    CompletableFuture<Void> acknowledgeAsync(List<MessageId> messageIdList, Transaction txn)
+            throws PulsarClientException;
```
Comment (Technoboy-) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java:537]: And here.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java [hunk 6]
```diff
@@ -492,11 +492,30 @@ public CompletableFuture<Void> acknowledgeAsync(Messages<?> messages) {
...
+    }
+
     @Override
     public CompletableFuture<Void> acknowledgeAsync(List<MessageId> messageIdList) {
         return doAcknowledgeWithTxn(messageIdList, AckType.Individual, Collections.emptyMap(), null);
     }
 
+    @Override
+    public CompletableFuture<Void> acknowledgeAsync(List<MessageId> messageIdList, Transaction txn) {
+        return doAcknowledgeWithTxn(messageIdList, AckType.Individual, Collections.emptyMap(), (TransactionImpl) txn);
```
Comment (congbobo184) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:516]: should register the partition topic and doAckForResponse

**Review by congbobo184 — CHANGES_REQUESTED** [review]

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java [hunk 7]
```diff
@@ -414,13 +414,6 @@ void reconsumeLater(Message<?> message,
      * @param messageId
      *            The {@code MessageId} to be cumulatively acknowledged
      * @param txn {@link Transaction} the transaction to cumulative ack
-     * @throws PulsarClientException.AlreadyClosedException
-     *             if the consumer was already closed
-     * @throws org.apache.pulsar.client.api.PulsarClientException.TransactionConflictException
-     *             if the ack with messageId is less than the messageId in pending ack state or ack with transaction is
-     *             different from the transaction in pending ack.
-     * @throws org.apache.pulsar.client.api.PulsarClientException.NotAllowedException
-     *             broker don't support transaction
```
Comment (congbobo184) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java:423]: don't delete this

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java [hunk 8]
```diff
@@ -469,13 +462,6 @@ CompletableFuture<Void> acknowledgeCumulativeAsync(MessageId messageId,
      *
      * @param messageId {@link MessageId} to be individual acknowledged
      * @param txn {@link Transaction} the transaction to cumulative ack
-     * @throws PulsarClientException.AlreadyClosedException
-     *             if the consumer was already closed
-     * @throws org.apache.pulsar.client.api.PulsarClientException.TransactionConflictException
-     *             if the ack with messageId has been acked by another transaction
-     * @throws org.apache.pulsar.client.api.PulsarClientException.NotAllowedException
-     *             broker don't support transaction
-     *             don't find batch size in consumer pending ack
```
Comment (congbobo184) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java:478]: if you want to delete this, please add the sync method in the consumer API

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java [hunk 9]
```diff
@@ -606,12 +625,21 @@ protected CompletableFuture<Void> doAcknowledgeWithTxn(List<MessageId> messageId
...
+
+            // it is okay that we register acked topic after sending the acknowledgements. because
+            // the transactional ack will not be visiable for consumers until the transaction is
+            // committed
+            if (ackType == AckType.Cumulative) {
+                txn.registerCumulativeAckConsumer((ConsumerImpl<?>) this);
+            }
+
             ackFuture = txn.registerAckedTopic(getTopic(), subscription)
                     .thenCompose(ignored -> doAcknowledge(messageIdList, ackType, properties, txn));
```
Comment (congbobo184) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:638]: we don't need regist ack in this code

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 10]
```diff
@@ -578,11 +579,32 @@ protected CompletableFuture<Void> doAcknowledge(MessageId messageId, AckType ack
...
+                onAcknowledge(messageIdList, exception);
+            } else if (AckType.Cumulative.equals(ackType)) {
+                onAcknowledgeCumulative(messageIdList, exception);
+            }
+            return FutureUtil.failedFuture(exception);
+        }
+        if (txn != null) {
+            List<CompletableFuture<Void>> completableFutures = new LinkedList<>();
+            for (MessageId messageId : messageIdList) {
+                completableFutures.add(doTransactionAcknowledgeForResponse(messageId, ackType, null,
```
Comment (congbobo184) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:601]: the proto support send messgeIdList, so you should rewrite a command is better.
https://github.com/apache/pulsar/blob/9b01d1c472aa5cde5205ad86fa3b8f998da95f35/pulsar-common/src/main/proto/PulsarApi.proto#L553

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 11]
```diff
@@ -2682,6 +2700,58 @@ private CompletableFuture<Void> doTransactionAcknowledgeForResponse(MessageId me
...
+                    bitSetRecyclable.set(0, batchMessageId.getBatchSize());
+                    bitSetRecyclable.clear(batchMessageId.getBatchIndex());
+                }
+                MessageIdData messageIdData = new MessageIdData();
+                messageIdData.setLedgerId(batchMessageId.getLedgerId());
+                messageIdData.setEntryId(batchMessageId.getEntryId());
+                messageIdData.setBatchSize(batchMessageId.getBatchSize());
+                long[] as = bitSetRecyclable.toLongArray();
+                for (int i = 0; i < as.length; i++) {
+                    messageIdData.addAckSet(as[i]);
```
Comment (congbobo184) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2730]: same as

```
private CompletableFuture<Void> doTransactionAcknowledgeForResponse(MessageId messageId, AckType ackType,
                                                                        ValidationError validationError,
                                                                        Map<String, Long> properties, TxnID txnID) {
```

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 12]
```diff
@@ -2699,6 +2717,58 @@ private CompletableFuture<Void> doTransactionAcknowledgeForResponse(MessageId me
...
+        for (MessageId messageId : messageIds) {
+            if (messageId instanceof BatchMessageIdImpl) {
+                BatchMessageIdImpl batchMessageId = (BatchMessageIdImpl) messageId;
+                bitSetRecyclable = BitSetRecyclable.create();
+                if (ackType == AckType.Cumulative) {
+                    batchMessageId.ackCumulative();
+                    bitSetRecyclable.set(0, batchMessageId.getBatchSize());
+                    bitSetRecyclable.clear(0, batchMessageId.getBatchIndex() + 1);
+                } else {
+                    bitSetRecyclable.set(0, batchMessageId.getBatchSize());
```
Comment (congbobo184) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2738]: same as
```
private CompletableFuture<Void> doTransactionAcknowledgeForResponse(MessageId messageId, AckType ackType,
                                                                        ValidationError validationError,
                                                                        Map<String, Long> properties, TxnID txnID) {
```

**Review by congbobo184 — APPROVED** [review]
LGTM! Please optimize some repetitive code

**Review by eolivelli — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 13]
```diff
@@ -2704,6 +2722,58 @@ private CompletableFuture<Void> doTransactionAcknowledgeForResponse(MessageId me
...
+    private CompletableFuture<Void> doTransactionAcknowledgeForResponse(List<MessageId> messageIds, AckType ackType,
+                                                                        ValidationError validationError,
+                                                                        Map<String, Long> properties, TxnID txnID) {
+        BitSetRecyclable bitSetRecyclable = null;
+        long ledgerId;
+        long entryId;
+        ByteBuf cmd;
+        long requestId = client.newRequestId();
+        List<MessageIdData> messageIdDataList = new LinkedList<>();
+        for (MessageId messageId : messageIds) {
```
Comment (gaoran10) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2734]: Do we need to check if existing ack position conflication? It seems that the broker also will check conflication, maybe we could add conflication check at client.

Comment (liangyepianzhou) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2734]: Yeah, the broker will check confliction.

**Pull request merged**